In [1]:
# load initial dependencies
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
# load datasets
control_group = pd.read_csv("control_group.csv",sep = ";")
test_group = pd.read_csv("test_group.csv",sep = ";")

In [5]:
# lets see few records
control_group.head()

Campaign Name       Date  Spend [USD]  # of Impressions     Reach  \
0  Control Campaign  1.08.2019         2280           82702.0   56930.0   
1  Control Campaign  2.08.2019         1757          121040.0  102513.0   
2  Control Campaign  3.08.2019         2343          131711.0  110862.0   
3  Control Campaign  4.08.2019         1940           72878.0   61235.0   
4  Control Campaign  5.08.2019         1835               NaN       NaN   

   # of Website Clicks  # of Searches  # of View Content  # of Add to Cart  \
0               7016.0         2290.0             2159.0            1819.0   
1               8110.0         2033.0             1841.0            1219.0   
2               6508.0         1737.0             1549.0            1134.0   
3               3065.0         1042.0              982.0            1183.0   
4                  NaN            NaN                NaN               NaN   

   # of Purchase  
0          618.0  
1          511.0  
2          372.0  
3          340.0  
4            NaN

In [6]:
# lets see few records
test_group.head()

Campaign Name       Date  Spend [USD]  # of Impressions  Reach  \
0  Test Campaign  1.08.2019         3008             39550  35820   
1  Test Campaign  2.08.2019         2542            100719  91236   
2  Test Campaign  3.08.2019         2365             70263  45198   
3  Test Campaign  4.08.2019         2710             78451  25937   
4  Test Campaign  5.08.2019         2297            114295  95138   

   # of Website Clicks  # of Searches  # of View Content  # of Add to Cart  \
0                 3038           1946               1069               894   
1                 4657           2359               1548               879   
2                 7885           2572               2367              1268   
3                 4216           2216               1437               566   
4                 5863           2106                858               956   

   # of Purchase  
0            255  
1            677  
2            578  
3            340  
4            768

In [7]:
# general information
control_group.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Campaign Name        30 non-null     object 
 1   Date                 30 non-null     object 
 2   Spend [USD]          30 non-null     int64  
 3   # of Impressions     29 non-null     float64
 4   Reach                29 non-null     float64
 5   # of Website Clicks  29 non-null     float64
 6   # of Searches        29 non-null     float64
 7   # of View Content    29 non-null     float64
 8   # of Add to Cart     29 non-null     float64
 9   # of Purchase        29 non-null     float64
dtypes: float64(7), int64(1), object(2)
memory usage: 2.5+ KB


In [8]:
# general information
test_group.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30 entries, 0 to 29
Data columns (total 10 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Campaign Name        30 non-null     object
 1   Date                 30 non-null     object
 2   Spend [USD]          30 non-null     int64 
 3   # of Impressions     30 non-null     int64 
 4   Reach                30 non-null     int64 
 5   # of Website Clicks  30 non-null     int64 
 6   # of Searches        30 non-null     int64 
 7   # of View Content    30 non-null     int64 
 8   # of Add to Cart     30 non-null     int64 
 9   # of Purchase        30 non-null     int64 
dtypes: int64(8), object(2)
memory usage: 2.5+ KB


In [9]:
# check null values percentage
control_group.isna().sum()/len(control_group)*100

Campaign Name          0.000000
Date                   0.000000
Spend [USD]            0.000000
# of Impressions       3.333333
Reach                  3.333333
# of Website Clicks    3.333333
# of Searches          3.333333
# of View Content      3.333333
# of Add to Cart       3.333333
# of Purchase          3.333333
dtype: float64

In [10]:
# lets handle it by imputation 
control_group[['# of Impressions', 'Reach',
       '# of Website Clicks', '# of Searches', '# of View Content',
       '# of Add to Cart', '# of Purchase']] = control_group[['# of Impressions', 'Reach',
       '# of Website Clicks', '# of Searches', '# of View Content',
       '# of Add to Cart', '# of Purchase']].mean()

In [11]:
# check null values percentage again
control_group.isna().sum()/len(control_group)*100

Campaign Name          0.0
Date                   0.0
Spend [USD]            0.0
# of Impressions       0.0
Reach                  0.0
# of Website Clicks    0.0
# of Searches          0.0
# of View Content      0.0
# of Add to Cart       0.0
# of Purchase          0.0
dtype: float64

In [13]:
# statistical information
control_group.describe()

Spend [USD]  # of Impressions         Reach  # of Website Clicks  \
count    30.000000         30.000000  3.000000e+01         3.000000e+01   
mean   2288.433333     109559.758621  8.884493e+04         5.320793e+03   
std     367.334451          0.000000  4.440205e-11         2.775128e-12   
min    1757.000000     109559.758621  8.884493e+04         5.320793e+03   
25%    1945.500000     109559.758621  8.884493e+04         5.320793e+03   
50%    2299.500000     109559.758621  8.884493e+04         5.320793e+03   
75%    2532.000000     109559.758621  8.884493e+04         5.320793e+03   
max    3083.000000     109559.758621  8.884493e+04         5.320793e+03   

       # of Searches  # of View Content  # of Add to Cart  # of Purchase  
count      30.000000          30.000000              30.0   3.000000e+01  
mean     2221.310345        1943.793103            1300.0   5.227931e+02  
std         0.000000           0.000000               0.0   1.156303e-13  
min      2221.310345        1943.793103            1300.0   5.227931e+02  
25%      2221.310345        1943.793103            1300.0   5.227931e+02  
50%      2221.310345        1943.793103            1300.0   5.227931e+02  
75%      2221.310345        1943.793103            1300.0   5.227931e+02  
max      2221.310345        1943.793103            1300.0   5.227931e+02

In [14]:
# statistical information
test_group.describe()

Spend [USD]  # of Impressions          Reach  # of Website Clicks  \
count    30.000000         30.000000      30.000000            30.000000   
mean   2563.066667      74584.800000   53491.566667          6032.333333   
std     348.687681      32121.377422   28795.775752          1708.567263   
min    1968.000000      22521.000000   10598.000000          3038.000000   
25%    2324.500000      47541.250000   31516.250000          4407.000000   
50%    2584.000000      68853.500000   44219.500000          6242.500000   
75%    2836.250000      99500.000000   78778.750000          7604.750000   
max    3112.000000     133771.000000  109834.000000          8264.000000   

       # of Searches  # of View Content  # of Add to Cart  # of Purchase  
count      30.000000          30.000000         30.000000      30.000000  
mean     2418.966667        1858.000000        881.533333     521.233333  
std       388.742312         597.654669        347.584248     211.047745  
min      1854.000000         858.000000        278.000000     238.000000  
25%      2043.000000        1320.000000        582.500000     298.000000  
50%      2395.500000        1881.000000        974.000000     500.000000  
75%      2801.250000        2412.000000       1148.500000     701.000000  
max      2978.000000        2801.000000       1391.000000     890.000000

In [15]:
# lets confirm unique values
print(control_group['Campaign Name'].value_counts())

print(test_group['Campaign Name'].value_counts())

Campaign Name
Control Campaign    30
Name: count, dtype: int64
Campaign Name
Test Campaign    30
Name: count, dtype: int64


In [16]:
# calculate mean metric
control_purchases = control_group['# of Purchase']
test_purchases = test_group['# of Purchase']

statistical significance

In [17]:
from scipy import stats

# perform statistical testing | t-test
# two-sample t-test to compare the means of the two groups.
# The null hypothesis (H0) is that there's no significant difference between the groups.
# The alternative hypothesis (H1) is that the test group performs significantly better.
t_stat, p_value = stats.ttest_ind(control_purchases, test_purchases)
print("p value : {}".format(p_value))

if p_value < 0.05:  # You can adjust the significance level (alpha) as needed
    print("The Test Campaign significantly outperforms the Control Campaign.")
else:
    print("There is no significant difference between the Test and Control Campaigns.")

p value : 0.9678494754256669
There is no significant difference between the Test and Control Campaigns.


C:\Users\Bidisha Shit\AppData\Local\Programs\Python\Python310\lib\site-packages\scipy\stats\_axis_nan_policy.py:523: RuntimeWarning: Precision loss occurred in moment calculation due to catastrophic cancellation. This occurs when the data are nearly identical. Results may be unreliable.
  res = hypotest_fun_out(*samples, **kwds)


interpret results:


The p-value is a measure of the evidence against a null hypothesis. In A/B testing, the null hypothesis often states that there is no significant difference between the two groups (in this case, the Test and Control Campaigns). The p-value tells you how likely it is to observe a difference as extreme as the one you've observed in your data, assuming that the null hypothesis is true.
A high p-value (close to 1) suggests that the observed difference between the two groups is likely due to random chance rather than a real, significant effect. In your case, a p-value of 0.9678 is very close to 1, indicating a high probability that any observed differences in the number of purchases between the Test and Control Campaigns are not statistically significant.

conclusion


Based on the p-value and the common significance level (alpha) of 0.05, which is often used in A/B testing, we fail to reject the null hypothesis. In other words, there is no strong statistical evidence to suggest that the Test Campaign is performing significantly better or worse than the Control Campaign in terms of the number of purchases.
Given this result, you might conclude that, for the metric of purchases, there is no compelling reason to prefer the Test Campaign over the Control Campaign. Both campaigns seem to have a similar impact on the number of purchases, at least based on the data you've analyzed

In [18]:
# calculate the difference in practical significance metrics between Test and Control Campaigns.
# these metrics could include conversion rates, revenue, customer retention, etc.
control_conversion_rate = control_group['# of Purchase'].sum() / control_group['# of Website Clicks'].sum()
test_conversion_rate = test_group['# of Purchase'].sum() / test_group['# of Website Clicks'].sum()

# calculate the absolute difference in conversion rates.
conversion_rate_difference = test_conversion_rate - control_conversion_rate

# you can set a practical significance threshold based on business goals.
practical_significance_threshold = 0.02 
if abs(conversion_rate_difference) >= practical_significance_threshold:
    print("There is a practical significance difference between the Test and Control Campaigns.")
else:
    print("There is no practical significance difference between the Test and Control Campaigns.")

There is no practical significance difference between the Test and Control Campaigns.


conclusion


it suggests that, from a practical standpoint, there is no meaningful difference between the two campaigns in terms of the chosen practical significance metric (e.g., conversion rate, revenue, or another relevant metric).

cost effectiveness

In [19]:
control_cost = control_group['Spend [USD]'].sum()
test_cost = test_group['Spend [USD]'].sum()

if test_cost < control_cost:
    print("The Test Campaign is more cost-effective.")
else:
    print("The Control Campaign is more cost-effective or equally cost-effective.")

The Control Campaign is more cost-effective or equally cost-effective.
